app/src/main/java/com/asma/tasky/feature_authentication/presentation/login/LoginViewModel.kt

```diff
@@ -0,0 +1,89 @@
+package com.asma.tasky.feature_authentication.presentation.login
+
+import androidx.compose.runtime.State
+import androidx.compose.runtime.mutableStateOf
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.asma.tasky.feature_authentication.domain.use_case.LoginUseCase
+import com.asma.tasky.core.util.Resource
+import com.asma.tasky.core.domain.states.TextFieldState
+import com.asma.tasky.core.presentation.util.UiEvent
+import com.asma.tasky.core.util.UiText
+import dagger.hilt.android.lifecycle.HiltViewModel
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.flow.asSharedFlow
+import kotlinx.coroutines.launch
+import javax.inject.Inject
+
+@HiltViewModel
+class LoginViewModel @Inject constructor(
+    private val loginUseCase: LoginUseCase
+) : ViewModel() {
+
+    private val _emailState = mutableStateOf(TextFieldState())
+    val emailState: State<TextFieldState> = _emailState
+
+    private val _passwordState = mutableStateOf(TextFieldState())
+    val passwordState: State<TextFieldState> = _passwordState
+
+    private val _loginState = mutableStateOf(LoginState())
+    val loginState: State<LoginState> = _loginState
+
+    private val _eventFlow = MutableSharedFlow<UiEvent>()
+    val eventFlow = _eventFlow.asSharedFlow()
+
+    fun onEvent(event: LoginEvent) {
+
+        when (event) {
+            is LoginEvent.EmailEntered -> {
+                _emailState.value = emailState.value.copy(
+                    text = event.email
+                )
```

Comment:
In general, I'd always use `_stateFlow.update { it.copy(...) }` to avoid race conditions. This is how it could happen:

*StateFlow holds value A*
- Thread 1 calls `copy()` on StateFlow with value A
- Before Thread 1 assigns its changed copy, Thread 2 updates StateFlow with value B
- Now, Thread 1 assigns its changed copy, which is still based on value A though, but the latest value is already value B. Therefore, value B would never be considered here.

Not always an issue, but I worked on highly stateful apps with frequently changing states, in which this really bit me, since such bugs are super hard to find.